### PR TITLE
Cli command to download artifacts from a URI. Updated R apis for loading / scoring models.

### PR DESCRIPTION
**Pull request by tomasatdatabricks**
## What changes are proposed in this pull request?
Update R apis for loading and serving models to use absolute URIs instead of path with optional run-id.
This is to adapt R api to the new MLflow convention of accessing artifacts from outside of mlflow.tracking component. 
 
## How is this patch tested?
- cli test in python (tests/store/test_cli.py#test_download_artifacts_from_uri)
Test that the new clie can download artifacts from run and file uri.
- integration test in R
Test that mlflow_load_model and mlflow_rfunc_predict now work with runs or actual uris.
 
## Release Notes
MLflow/R apis for loading and serving models (mlflow_load_model, mlflow_rfunc_serve, mlflow_rfunc_predict) now take a fully qualified URI instead of run-id and run-relative artifacts path. The new apis allow loading models and other artifacts from different sources in a consistent way without the requirement of each artifact to be referenced by an MLflow run.

### Is this a user-facing change? 

- [ ] No. You can skip the rest of this section.
- [ x] Yes. Give a description of this change to be included in the release notes for MLflow users.
 
(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)
 
### What component(s) does this PR affect?
 
- [ ] UI
- [ x] CLI 
- [ ] API 
- [ ] REST-API 
- [ ] Examples 
- [ ] Docs
- [ ] Tracking
- [ ] Projects 
- [ ] Artifacts 
- [ ] Models 
- [ ] Scoring 
- [ ] Serving
- [ x] R
- [ ] Java
- [ ] Python

### How should the PR be classified in the release notes? Choose one:
 
- [x ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/R/mlflow/R/model.R [hunk 1]
```diff
@@ -76,15 +76,19 @@ mlflow_predict_model <- function(model, data) {
 #' Loads an MLflow model. MLflow models can have multiple model flavors. Not all flavors / models
 #' can be loaded in R. This method by default searches for a flavor supported by R/MLflow.
 #'
-#' @param model_path Path to the MLflow model. The path is relative to the run with the given
-#'        run-id or local filesystem path without run-id.
-#' @param run_id Optional MLflow run-id. If supplied model will be fetched from MLflow tracking
-#'        server.
+#' @param model_uri The location, in URI format, of the MLflow model, for example:
+#'                  - ``file:///Users/me/path/to/local/model``
+#'                  - ``file:relative/path/to/local/model``
```
Comment (sueann) [inline mlflow/R/mlflow/R/model.R]: `file:/` ? Whether `file:` should work is ambiguous based on the definition of URIs (so it depends on the implementation). The docs say to use a slash so we might as well be consistent (https://www.mlflow.org/docs/latest/tracking.html#where-runs-are-recorded)

File: mlflow/R/mlflow/R/model.R [hunk 2]
```diff
@@ -146,23 +149,22 @@ mlflow_load_model <- function(model_path, flavor = NULL, run_id = NULL) {
 #' jsonlite::write_json(iris, "iris.json")
 #'
 #' # predict existing model from json data
-#' mlflow_rfunc_predict("mlflow_roundtrip", "iris.json")
+#' mlflow_rfunc_predict("file:mlflow_roundtrip", "iris.json")
```
Comment (sueann) [inline mlflow/R/mlflow/R/model.R]: `file:/`

File: mlflow/R/mlflow/R/model-serve.R [hunk 3]
```diff
@@ -4,8 +4,7 @@
 #'
 #' Serves an RFunc MLflow model as a local web API.
 #'
-#' @param model_path The path to the MLflow model, as a string.
-#' @param run_id ID of run to grab the model from.
+#' @param model_uri The URI to the MLflow model, as a string.
```
Comment (sueann) [inline mlflow/R/mlflow/R/model-serve.R]: I don't think we have to copy the list of URI examples for every docstring, but I think it's worth at least putting 
```
#                  For information about supported URI schemes, see the
#                  `Artifacts Documentation
#                  <https://www.mlflow.org/docs/latest/tracking.html#supported-artifact-stores>`_.
```

File: mlflow/R/mlflow/R/model.R [hunk 4]
```diff
@@ -127,8 +131,7 @@ mlflow_load_model <- function(model_path, flavor = NULL, run_id = NULL) {
 #'
 #' Performs prediction using an RFunc MLflow model from a file or data frame.
 #'
-#' @param model_path The path to the MLflow model, as a string.
-#' @param run_id Run ID of run to grab the model from.
+#' @param model_uri The URI to the MLflow model, as a string.
```
Comment (sueann) [inline mlflow/R/mlflow/R/model.R]:
```
#                  For information about supported URI schemes, see the
#                  `Artifacts Documentation
#                  <https://www.mlflow.org/docs/latest/tracking.html#supported-artifact-stores>`_.
```

File: mlflow/store/cli.py [hunk 5]
```diff
@@ -103,3 +103,14 @@ def download_artifacts(run_id, artifact_path):
     artifact_repo = get_artifact_repository(artifact_uri)
     artifact_location = artifact_repo.download_artifacts(artifact_path)
     print(artifact_location)
+
+
+@commands.command("download-from-uri")
+@click.option("--artifact-uri", "-a", required=True,
+              help="URI pointing to the artifact file or artifacts directory.")
+def download_artifacts_from_uri(artifact_uri):
```
Comment (sueann) [inline mlflow/store/cli.py]: i think this is okay but since we might change the main user function for downloading from URIs (to possibly `data.download_uri`, what do you think about making it experimental? (i'll do it in the marking-experimental PR)

**Review by sueann — COMMENTED** [review]
lgtm! just a few nits (and the R test failure I didn't look too deeply into) - thanks!

File: mlflow/store/cli.py (hunk 5 above)
Comment (dbczumar) [inline mlflow/store/cli.py]: What is the difference between this method and https://github.com/mlflow/mlflow/blob/eb51e79d190afb8d0dd5f987cd4be00b4dd744d2/mlflow/data.py#L85, which is already exposed as a CLI command via `mlflow download`?

Also, how does this relate to `mlflow artifacts download`, which should probably be repurposed to operate on absolute URIs (either restricted to URIs with the `runs:/` scheme or operating on the full set of URIs supported by the MLflow artifact repo implementation)

**Review by dbczumar — CHANGES_REQUESTED** [review]
@tomasatdatabricks I think we may be duplicating CLI entrypoints here. Please let me know if I'm mistaken.

File: mlflow/store/cli.py (hunk 5 above)
Comment (sueann) [reply mlflow/store/cli.py]: `mlflow artifacts download` requires the `run_id` and seems to be part of a set of [CLI commands](https://github.com/mlflow/mlflow/blob/eb51e79d190afb8d0dd5f987cd4be00b4dd744d2/mlflow/store/cli.py#L90) that also require `run_id`. I assumed those were meant to be used within the context of a tracking server. 

`data.download` is what we'd like to use ideally, and it'd make sense for it to do basically what `_download_artifacts_from_uri` does. @andrewmchen was looking at this but we didn't want to block on it for the runs:/ URI work as it may not make it into 1.0. (It's not completely straight-forward because there is a small difference between its behavior and that of `_download_artifacts_from_uri` (in appending the basename to the destination path), and we need to make sure projects don't break after the change).

in summary, to use either of those would require behavior changes for them, so i'd vote for not blocking on that and using the new CLI command, and marking the new command as experimental so we can swap it out as soon as we can with hopefully `mlflow download`.

File: mlflow/R/mlflow/R/model.R (hunk 1 above)
Comment (tomasatdatabricks) [reply mlflow/R/mlflow/R/model.R]: Hmm I read that if you want to have a relative path, you should use file:. I'll look up the source and put more info here.

File: mlflow/R/mlflow/R/model-serve.R (hunk 3 above)
Comment (tomasatdatabricks) [reply mlflow/R/mlflow/R/model-serve.R]: Good point, although it's a bit verbose to include everywhere. Would it make sense to just say <mlflow-artifact-uri> and expect users are able to search the docs?

I am ok copying that here as well, just in general I don't like if there is too much text in the comment.

Comment (kevinykuo) [reply mlflow/R/mlflow/R/model-serve.R]: I don't think it'd hurt to include the information everywhere (preferably in the R doc so people won't have to follow a link). Also since `model_uri` has the same doc across functions we can consider [inheriting param](http://r-pkgs.had.co.nz/man.html) docs or using a template (e.g. https://github.com/mlflow/mlflow/blob/0cd28559447fd3a5d993aca66263060373183493/mlflow/R/mlflow/R/tracking-experiments.R#L8)

File: mlflow/store/cli.py (hunk 5 above)
Comment (dbczumar) [reply mlflow/store/cli.py]: As discussed offline, I'm okay with adding this additional CLI method, pending follow-up PRs to remove the `mlflow download` CLI and, *potentially*, merge this with `mlflow artifacts download`.

**Review by dbczumar — APPROVED** [review]
LGTM from a CLI coherency perspective

File: mlflow/R/mlflow/R/model.R (hunk 1 above)
Comment (tomasatdatabricks) [reply mlflow/R/mlflow/R/model.R]: Talked offline. "file:" is the right scheme to use for relative paths. @sueann Do you think the comment is enough as is (it points file:relative/path/....) or do you think we should make it more explicit?

File: mlflow/store/cli.py (hunk 5 above)
Comment (tomasatdatabricks) [reply mlflow/store/cli.py]: I am ok marking this experimental. What do I do to mark this as such? :)

File: mlflow/R/mlflow/R/model.R (hunk 2 above)
Comment (tomasatdatabricks) [reply mlflow/R/mlflow/R/model.R]: Talked offline, "file:" is the correct prefix for relative paths.

File: mlflow/R/mlflow/R/model-serve.R (hunk 3 above)
Comment (tomasatdatabricks) [reply mlflow/R/mlflow/R/model-serve.R]: I've updated these to use the template argument, thanks for the suggestion!

File: mlflow/R/mlflow/man-roxygen/roxlate-model-uri.R [hunk 6]
```diff
@@ -0,0 +1,11 @@
+#' @param model_uri The location, in URI format, of the MLflow model.
+#' @family artifact uri
+#' @details The URI scheme must be supported by MLflow - i.e. there has to be an MLflow artifact
+#'          repository corresponding to the scheme of the URI. The content is expected to point to a
+#'          directory containing MLmodel. The following are examples of valid model uris:
+#'                  - ``file:///Users/me/path/to/local/model``
```
Comment (sueann) [inline mlflow/R/mlflow/man-roxygen/roxlate-model-uri.R]: super nit: ``file:///absolute/path/to/local/model`` might make the relative one more clear.

**Review by sueann — APPROVED** [review]
lgtm! just one super nit - feel free to ignore if you don't need to make any other updates to the PR. thanks!

**Review by kevinykuo — APPROVED** [review]
LGTM!

File: mlflow/R/mlflow/man-roxygen/roxlate-model-uri.R (hunk 6 above)
Comment (tomasatdatabricks) [reply mlflow/R/mlflow/man-roxygen/roxlate-model-uri.R]: Good point, I'll update it.
